File: Assignment_1/Q2_a.cpp
//C++ code to find the Matrix Multiplication of two matrices with floating point numbers.
//Speed up is calculated keeping program (1.)using recursion as the baseline
//Speed up of program x over program y = time taken by y/ time taken by x



#include <iostream>
#include <random>
#include <iomanip>
using namespace std;

int main()
{
    // Creating random double values
    double lower_bound = 0;
    double upper_bound = 1000;

    uniform_real_distribution<double> unif(lower_bound, upper_bound);
    default_random_engine re;


    // Initialization
    int N = 256; //{32, 64, 128, 256}
    float A[N][N], B[N][N], C[N][N];

    // Assigning elements to the first matrix A

    for(int i = 0; i < N; ++i)
        for(int j = 0; j < N; ++j)
        {
            A[i][j] = unif(re);
        }

    // Assigning elements to the second matrix B

    for(int i = 0; i < N; ++i)
        for(int j = 0; j < N; ++j)
        {
            B[i][j] = unif(re);
        }

    // Initializing the elements of matrix C to 0
    
    for(int i = 0; i < N; ++i)
        for(int j = 0; j < N; ++j)
        {
            C[i][j] = 0;
        }

    // Defining struct timespec ts that stores the time elapsed
    struct timespec ts;
    char buff[100];

    // Start Time
    timespec_get(&ts, TIME_UTC);
    unsigned int start = ts.tv_nsec;
    unsigned int start_sec = ts.tv_sec;

    strftime(buff, sizeof buff, "%D %T", gmtime(&ts.tv_sec));
    cout << "Start time: "<< buff << "." << ts.tv_nsec << " UTC" << endl;


    // Multiplying matrix A and B and storing in array C.
    for(int i = 0; i < N; ++i)
        for(int j = 0; j < N; ++j)
            for(int k = 0; k < N; ++k)
            {
                C[i][j] += A[i][k] * B[k][j];
            }

    
    // End Time
    timespec_get(&ts, TIME_UTC);
    unsigned int end = ts.tv_nsec;
    unsigned int end_sec = ts.tv_sec;

    strftime(buff, sizeof buff, "%D %T", gmtime(&ts.tv_sec));
    
    // Displaying the multiplication of two matrix.

    for(int i = 0; i < N; ++i){
        for(int j = 0; j < N; ++j)
    {
        cout << setprecision(10) << " " << C[i][j];
        if(j == N-1)
            cout << endl;
    }
    }

    // Printing the end time for multiplication (Only meat portion of the code)
    cout << "End time: "<< buff << "." << ts.tv_nsec << " UTC" << endl;

    // Total Time Elapsed
    cout << "Execution Time (Matrix Multiplication): " << end_sec - start_sec << " s " << end - start << " ns" << endl;

    return 0;
}

File: Assignment_1/Q2_a.py
from __future__ import print_function
import random
import timeit


# Storing elements of a matrix.

def matrix_formation(N):
    A = [[0]*N]*N
    for i in range(N):
        for j in range(N):
            A[i][j] = round(random.uniform(0.0000, 999.9999), 4)

    return A

# Matrix Multiplication.

def Multiplication(N):
    start = timeit.default_timer()
    A = matrix_formation(N)
    B = matrix_formation(N)

    # Multiplying matrix A and B and storing in array mult.
    C = [[0]*N]*N
    for x in range(N):
        for y in range(N):
            for z in range(N):
                C[x][y] += A[x][z] + B[z][y]

    stop = timeit.default_timer()
    elapsed_time = stop - start
    return C, elapsed_time

# Printing resultant matrix.

def print_matrix(N):

    C, elapsed_time = Multiplication(N)
    for x in range(N):
        for y in range(N):
            print(C[x][y], end=' ')
            if(y == N-1):
                print('\n')

    print("Execution Time (Matrix Multiplication only): ", elapsed_time)

start_total = timeit.default_timer()

print_matrix(32)

stop_total = timeit.default_timer()

# Time Elapsed
print("Execution Time (Total): ", stop_total - start_total, " s")
